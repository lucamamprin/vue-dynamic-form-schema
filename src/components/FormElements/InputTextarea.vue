<template>
  <div
    class="input textarea floating-label"
    :class="{'input-transparent': darkMode, 'mt-0': resetMarginTop, 'floating-label-floated': typedText.length || floatedLabel}"
  >
    <div :class="areaClass">
      <textarea
        :id="id"
        :name="name"
        :disabled="disabled"
        :rows="rows"
        :cols="cols"
        :minlength="minLength"
        :maxlength="maxLength"
        :aria-describedby="`${describedBy} ${maxLengthDescribedBy}`"
        :aria-invalid="invalid.toString()"
        :aria-required="required.toString()"
        :aria-labelledby="`label-${name}`"
        :class="{'border-danger': hasError,}"
        :style="inputStyle"
        :placeholder="placeholderText"
        @input="onType($event)"
        @keyup.enter="resize"
        @keyup.delete="resize"
        v-model="typedText"
      />
      <label
        :for="id"
        v-html="label"
      />

      <textarea
        v-if="resizeTextArea"
        :id="`shadow-${name}`"
        ref="shadow"
        v-model="typedText"
        class="shadow"
        :style="shadowStyles"
        tabindex="-1"
        :disabled="disabled"
      />
      <label
        v-if="resizeTextArea"
        :for="`shadow-${name}`"
        class="sr-only"
      >{{ label }}</label>
    </div>

    <max-length
      v-if="hasCharacterCount"
      :typed-text="typedText"
      :max-length="maxLength"
      :sr-character-count="maxLengthDescribedBy"
      :label="removeStar(label)"
      classes="pr-lg-14"
    />

    <div
      v-if="hasError && error"
      class="mb-13"
    >
      <error-message
        :id="describedBy"
        :error-message="error"
      />
    </div>
  </div>
</template>

<script>
import { removeStar } from "../../utilities/removeStar"

const MaxLength = () => import("./Atoms/MaxLength")
const ErrorMessage = () => import("./Atoms/ErrorMessage")

export default {
  name: "InputTextarea",
  data () {
    return {
      valueProxy: "",
      inputHeight: "0",
      shadowStyles: {
        maxHeight: 0,
        pointerEvents: "none",
        opacity: 0,
        margin: 0,
        padding: 0,
      },
    }
  },
  components: {
    MaxLength,
    ErrorMessage,
  },
  props: {
    // textarea specific props
    resizeTextArea: {
      type: Boolean,
      default: true,
    },
    rows: {
      type: Number,
      default: 2,
    },
    cols: {
      type: Number,
      default: 20,
    },
    areaClass: {
      type: String,
      default: "",
    },
    disabled: {
      type: Boolean,
      default: false,
    },

    // common with BaseInput
    floatedLabel: {
      type: Boolean,
      default: false,
    },
    id: {
      type: String,
      required: true,
    },
    describedBy: {
      type: String,
      required: true,
    },
    value: {
      type: [
        String,
        Number,
      ],
      default: "",
    },
    name: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      required: true,
    },
    error: {
      type: String,
      default: "",
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    minLength: {
      type: Number,
      default: 0,
    },
    maxLength: {
      type: Number,
      default: 0,
    },
    darkMode: {
      type: Boolean,
      default: false,
    },
    showCharacterCount: {
      type: Boolean,
      default: true,
    },
    placeholderText: {
      type: String,
      default: " ",
    },
    invalid: {
      type: Boolean,
    },
    required: {
      type: Boolean,
      default: false,
    },
    resetMarginTop: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    typedText: {
      get () {
        return this.value
      },
      set (newValue) {
        this.valueProxy = newValue
      },
    },
    inputStyle () {
      if (!this.resizeTextArea) {
        return
      }

      return {
        "min-height": this.inputHeight,
        overflow: "hidden",
      }
    },
    hasCharacterCount () {
      return this.maxLength && this.showCharacterCount
    },
    maxLengthDescribedBy () {
      return this.hasCharacterCount ? `${this.id}-counter` : ""
    },
  },
  methods: {
    removeStar,
    onType ($event) {
      const value = $event.target.value

      this.$emit("input", value)
      this.valueProxy = value
      this.resize()
    },
    resize () {
      // Credits to https://www.scottstadt.com/2019/06/03/vue-autosize-textarea.html
      if (!this.resizeTextArea) {
        return
      }

      this.inputHeight = `${this.$refs.shadow.scrollHeight}px`
    },
  },
  mounted () {
    this.resize()
  },
}
</script>
