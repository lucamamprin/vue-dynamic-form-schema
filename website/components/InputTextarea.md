# InputTextarea

## Props

<!-- @vuese:InputTextarea:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|resizeTextArea|textarea specific props|`Boolean`|`false`|true|
|rows|-|`Number`|`false`|2|
|cols|-|`Number`|`false`|20|
|areaClass|-|`String`|`false`|-|
|disabled|-|`Boolean`|`false`|false|
|floatedLabel|common with BaseInput|`Boolean`|`false`|false|
|id|-|`String`|`true`|-|
|describedBy|-|`String`|`true`|-|
|value|-|`String` /  `Number`|`false`|-|
|name|-|`String`|`true`|-|
|label|-|`String`|`true`|-|
|error|-|`String`|`false`|-|
|hasError|-|`Boolean`|`false`|false|
|minLength|-|`Number`|`false`|0|
|maxLength|-|`Number`|`false`|0|
|darkMode|-|`Boolean`|`false`|false|
|showCharacterCount|-|`Boolean`|`false`|true|
|placeholderText|-|`String`|`false`| |
|invalid|-|`Boolean`|`false`|-|
|required|-|`Boolean`|`false`|false|
|resetMarginTop|-|`Boolean`|`false`|false|

<!-- @vuese:InputTextarea:props:end -->


## Events

<!-- @vuese:InputTextarea:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|input|-|-|

<!-- @vuese:InputTextarea:events:end -->


## MixIns

<!-- @vuese:InputTextarea:mixIns:start -->
|MixIn|
|---|
|removeStar|

<!-- @vuese:InputTextarea:mixIns:end -->


