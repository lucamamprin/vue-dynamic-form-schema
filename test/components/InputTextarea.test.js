import Vue from "vue"
import { shallowMount } from "@vue/test-utils"
import InputTextarea from "../../src/components/FormElements/InputTextarea.vue"

const props = {
  id: "message",
  describedBy: "message-describe",
  name: "message",
  label: "Your Message",
}

let wrapper

describe("InputTextarea", () => {
  beforeEach(() => {
    wrapper = shallowMount(InputTextarea, {
      propsData: props,
    })
  })

  test("mounts correctly", () => {
    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test("show minlength and maxlength attributes and component", async () => {
    await wrapper.setProps({
      minLength: 0,
      maxLength: 500,
    })

    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  test("display value if already set", async () => {
    const value = "Lorem ipsum dolor sit amet"
    await wrapper.setProps({
      value,
    })

    await Vue.nextTick()

    expect(wrapper.find("textarea").element.value).toBe(value)
  })

  test("accepts different cols and rows", async () => {
    await wrapper.setProps({
      cols: 40,
      rows: 40,
    })

    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  test("accepts custom class", async () => {
    await wrapper.setProps({
      areaClass: "my-custom-class",
    })

    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  test("can be disabled", async () => {
    await wrapper.setProps({
      disabled: true,
    })

    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  test("shadow textarea won't show if resizing is disabled", async () => {
    await wrapper.setProps({
      resizeTextArea: false,
    })

    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })
})
