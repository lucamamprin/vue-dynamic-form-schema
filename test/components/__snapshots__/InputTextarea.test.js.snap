// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputTextarea accepts custom class 1`] = `
<div class="input textarea floating-label">
  <div class="my-custom-class"><textarea id="message" name="message" rows="2" cols="20" minlength="0" maxlength="0" aria-describedby="message-describe " aria-invalid="false" aria-required="false" aria-labelledby="label-message" placeholder=" " class="" style="min-height: 0px; overflow: hidden;"></textarea> <label for="message">Your Message</label> <textarea id="shadow-message" tabindex="-1" class="shadow" style="max-height: 0; pointer-events: none; opacity: 0; margin: 0px; padding: 0px;"></textarea> <label for="shadow-message" class="sr-only">Your Message</label></div>
  <!---->
  <!---->
</div>
`;

exports[`InputTextarea accepts different cols and rows 1`] = `
<div class="input textarea floating-label">
  <div class=""><textarea id="message" name="message" rows="40" cols="40" minlength="0" maxlength="0" aria-describedby="message-describe " aria-invalid="false" aria-required="false" aria-labelledby="label-message" placeholder=" " class="" style="min-height: 0px; overflow: hidden;"></textarea> <label for="message">Your Message</label> <textarea id="shadow-message" tabindex="-1" class="shadow" style="max-height: 0; pointer-events: none; opacity: 0; margin: 0px; padding: 0px;"></textarea> <label for="shadow-message" class="sr-only">Your Message</label></div>
  <!---->
  <!---->
</div>
`;

exports[`InputTextarea can be disabled 1`] = `
<div class="input textarea floating-label">
  <div class=""><textarea id="message" name="message" rows="2" cols="20" minlength="0" maxlength="0" aria-describedby="message-describe " aria-invalid="false" aria-required="false" aria-labelledby="label-message" placeholder=" " class="" style="min-height: 0px; overflow: hidden;" disabled="disabled"></textarea> <label for="message">Your Message</label> <textarea id="shadow-message" tabindex="-1" class="shadow" style="max-height: 0; pointer-events: none; opacity: 0; margin: 0px; padding: 0px;" disabled="disabled"></textarea> <label for="shadow-message" class="sr-only">Your Message</label></div>
  <!---->
  <!---->
</div>
`;

exports[`InputTextarea mounts correctly 1`] = `
<div class="input textarea floating-label">
  <div class=""><textarea id="message" name="message" rows="2" cols="20" minlength="0" maxlength="0" aria-describedby="message-describe " aria-invalid="false" aria-required="false" aria-labelledby="label-message" placeholder=" " class="" style="min-height: 0px; overflow: hidden;"></textarea> <label for="message">Your Message</label> <textarea id="shadow-message" tabindex="-1" class="shadow" style="max-height: 0; pointer-events: none; opacity: 0; margin: 0px; padding: 0px;"></textarea> <label for="shadow-message" class="sr-only">Your Message</label></div>
  <!---->
  <!---->
</div>
`;

exports[`InputTextarea shadow textarea won't show if resizing is disabled 1`] = `
<div class="input textarea floating-label">
  <div class=""><textarea id="message" name="message" rows="2" cols="20" minlength="0" maxlength="0" aria-describedby="message-describe " aria-invalid="false" aria-required="false" aria-labelledby="label-message" placeholder=" " class="" style=""></textarea> <label for="message">Your Message</label>
    <!---->
    <!---->
  </div>
  <!---->
  <!---->
</div>
`;

exports[`InputTextarea show minlength and maxlength attributes and component 1`] = `
<div class="input textarea floating-label">
  <div class=""><textarea id="message" name="message" rows="2" cols="20" minlength="0" maxlength="500" aria-describedby="message-describe message-counter" aria-invalid="false" aria-required="false" aria-labelledby="label-message" placeholder=" " class="" style="min-height: 0px; overflow: hidden;"></textarea> <label for="message">Your Message</label> <textarea id="shadow-message" tabindex="-1" class="shadow" style="max-height: 0; pointer-events: none; opacity: 0; margin: 0px; padding: 0px;"></textarea> <label for="shadow-message" class="sr-only">Your Message</label></div>
  <div class="text-right text-light" classes="pr-lg-14"><span id="message-counter" class="sr-only">
    0 characters on 500 available for Your Message
  </span>
    <p class="m-0">
      0 / 500
    </p>
  </div>
  <!---->
</div>
`;
